signature-aggregator/aggregator/aggregator.go

```diff
@@ -0,0 +1,506 @@
+		if err != nil {
+			msg := "Failed to aggregate signature."
+			s.logger.Error(
+				msg,
+				zap.String("sourceBlockchainID", unsignedMessage.SourceChainID.String()),
+				zap.String("warpMessageID", unsignedMessage.ID().String()),
+				zap.Error(err),
+			)
+			return nil, true, fmt.Errorf("%s: %w", msg, err)
+		}
+
+		signedMsg, err := avalancheWarp.NewMessage(
+			unsignedMessage,
+			&avalancheWarp.BitSetSignature{
+				Signers:   vdrBitSet.Bytes(),
+				Signature: *(*[bls.SignatureLen]byte)(bls.SignatureToBytes(aggSig)),
+			},
+		)
+		if err != nil {
+			msg := "Failed to create new signed message"
+			s.logger.Error(
+				msg,
+				zap.String("sourceBlockchainID", unsignedMessage.SourceChainID.String()),
+				zap.String("warpMessageID", unsignedMessage.ID().String()),
+				zap.Error(err),
+			)
+			return nil, true, fmt.Errorf("%s: %w", msg, err)
+		}
+		return signedMsg, true, nil
+	}
+	// Not enough signatures, continue processing messages
+	return nil, true, nil
+}
+
+// isValidSignatureResponse tries to generate a signature from the peer.AsyncResponse, then verifies
+// the signature against the node's public key. If we are unable to generate the signature or verify
+// correctly, false will be returned to indicate no valid signature was found in response.
+func (s *SignatureAggregator) isValidSignatureResponse(
+	unsignedMessage *avalancheWarp.UnsignedMessage,
+	response message.InboundMessage,
+	pubKey *bls.PublicKey,
+) (blsSignatureBuf, bool) {
+	// If the handler returned an error response, count the response and continue
+	if response.Op() == message.AppErrorOp {
+		s.logger.Debug(
+			"Relayer async response failed",
+			zap.String("nodeID", response.NodeID().String()),
+		)
+		return blsSignatureBuf{}, false
+	}
+
+	appResponse, ok := response.Message().(*p2p.AppResponse)
+	if !ok {
+		s.logger.Debug(
+			"Relayer async response was not an AppResponse",
+			zap.String("nodeID", response.NodeID().String()),
+		)
+		return blsSignatureBuf{}, false
+	}
+
+	var sigResponse msg.SignatureResponse
+	if _, err := msg.Codec.Unmarshal(appResponse.AppBytes, &sigResponse); err != nil {
+		s.logger.Error(
+			"Error unmarshaling signature response",
+			zap.Error(err),
+		)
+	}
+	signature := sigResponse.Signature
+
+	// If the node returned an empty signature, then it has not yet seen the warp message. Retry later.
+	emptySignature := blsSignatureBuf{}
+	if bytes.Equal(signature[:], emptySignature[:]) {
+		s.logger.Debug(
+			"Response contained an empty signature",
+			zap.String("nodeID", response.NodeID().String()),
+		)
+		return blsSignatureBuf{}, false
+	}
+
+	sig, err := bls.SignatureFromBytes(signature[:])
+	if err != nil {
+		s.logger.Debug(
+			"Failed to create signature from response",
+		)
+		return blsSignatureBuf{}, false
+	}
+
+	if !bls.Verify(pubKey, sig, unsignedMessage.Bytes()) {
+		s.logger.Debug(
+			"Failed verification for signature",
+		)
+		return blsSignatureBuf{}, false
+	}
+
+	return signature, true
```

Comment:
Would it make more sense to unmarshal the signature here instead of in `aggregateSignatures`?

Reply:
Potentially but we have a follow up ticket to re-evaluate this logic: Referenced this in #388